### DISCUSSION TITLE
Why this Binding to Observable does not work?

### DISCUSSION THREAD

**Discussion by SuperJMN** [AvaloniaUI/Avalonia#8641] [Q&A]
I have created a binding to an Observable that is pushing elements. However, Avalonia isn't reacting to the updates.

The code is as follows:

```csharp
public class MainWindowViewModel : ReactiveObject
{
    private readonly ReadOnlyObservableCollection<MyViewModel> myCollection;

    public MainWindowViewModel()
    {
        var changes = Observable.Range(1, 4)
            .Select(i => new MyViewModel(i))
            .ToObservableChangeSet()
            .ObserveOn(RxApp.MainThreadScheduler)
            .Publish();

        changes
            .ObserveOn(RxApp.MainThreadScheduler)
            .Bind(out myCollection)
            .Subscribe();

        IsSomethingSelected = changes
            .AutoRefresh(x => x.IsSelected)
            .ToCollection()
            .Select(x => x.Any(model => model.IsSelected));

        IsSomethingSelected.Subscribe(v => Debug.WriteLine($"Something selected: {v}"));

        changes.Connect();
    }

    public IObservable<bool> IsSomethingSelected { get; }

    public ReadOnlyObservableCollection<MyViewModel> MyCollection => myCollection;
}
```
The binding code in XAML:

```xml
<CheckBox DockPanel.Dock="Top" Content="Has Selection?" IsEnabled="False" IsChecked="{Binding IsSomethingSelected^}" />
```

If you put a breakpoint in `Debug.WriteLine($"Something selected: {v}")` it will be hit, that is the normal behavior. But Avalonia ignores any updates happening to the `IsSomethingSelected` observable.

Am I missing something?

Thanks!

If you want the full code, I have a minimal repro project here:
https://github.com/SuperJMN/Test-Avalonia-BindToObservable

- Clone it.
- Run it.
- Select some items and notice how the top checkbox.

**Comment by timunie** [comment]
For me, it is not only ignoring the changes. The CheckBox has `null` as value. So the Binding does not pick up at all for me. 

![image](https://user-images.githubusercontent.com/47110241/182328818-534e6139-4192-41e7-b177-396f6f554f1b.png)
*CheckBox indeterminate*

**Reply by SuperJMN** [reply]
Hey @timunie thanks for your time!

That's super weird, because the observable is emitting values. If you don't mind, run it again and look to the output window while you select and unselect the checkboxes. You should see Debug.WriteLine is hit, that means that new values are arriving.

**Reply by timunie** [reply]
I've seen this. It's just not picking up the binding at all. So I think this may be a bug in the Binding engine.

**Comment by maxkatz6** [comment]
@SuperJMN I tried to find any problem in Avalonia, but there is nothing obvious for me so far.

I am not an expert in ReactiveUI (even though I am using it for like 2 years now haha), but I probably see what's wrong with this code you provided.
1. Debug.WriteLine subscription works because it's created BEFORE Connect call
2. Avalonia subscription doesn't work because it's created AFTER Connect call.

If I move `IsSomethingSelected.Subscribe(v => Debug.WriteLine($"Something selected: {v}"));` after Connect call it stops working even without Avalonia.

**Reply by maxkatz6** [reply]
Side note, I found out that Avalonia uses reflection to cast `IObservable<bool>` (or any other value type parameter) to `IObservable<object>`. Which means observables with value type generics probably aren't so optimal and might be problematic with AOT.
Though, it wasn't a reason of this bug.

**Reply by SuperJMN** [reply]
> Debug.WriteLine subscription works because it's created BEFORE Connect call
Avalonia subscription doesn't work because it's created AFTER Connect call.

I'm almost sure that Avalonia subscribes after the `changes.Connect()` call, because Avalonia uses the property after the constructor has been called, including that line.

**Reply by grokys** [reply]
Yes, that's what Max is saying. Avalonia subscribes after `changes.Connect()`, and `changes.Connect()` breaks the observable. You can see this by moving the `IsSomethingSelected.Subscribe(v => Debug.WriteLine($"Something selected: {v}"));` line to after `changes.Connect()`.

**Comment by timunie** [comment]
@SuperJMN I have looked into it again and provided an alternative for you: 

https://github.com/SuperJMN/Test-Avalonia-BindToObservable/pull/1

Note: I use MVVM Toolkit, because I find it has many benefits. You could also use only Reactive UI properties. Also it works in combination with ReactiveUI. 

Edit: Here are the docs: https://docs.microsoft.com/en-us/windows/communitytoolkit/mvvm/introduction

**Reply by SuperJMN** [reply]
Thanks @timunie But I'd like to use observables as much as I can. Mutability makes me feel dirty, haha :) It looks like that framework isn't designed to work well with reactive programming.

Anyways, I'll take a look to your PR today. It's always good to know how other alternatives look like!

**Reply by timunie** [reply]
> Mutability makes me feel dirty

I'm interested in the whys, if you can explain it to me. What are the benefits by using them? If it's just a matter of taste, I'm completely fine with it.

**Reply by SuperJMN** [reply]
> > Mutability makes me feel dirty
> 
> I'm interested in the whys, if you can explain it to me. What are the benefits by using them? If it's just a matter of taste, I'm completely fine with it.

As soon as you introduce degrees of immutability in a system, its complexity rises because you need to guard invariants carefully after each change. Look at this class:

```csharp
class Money
{
    public decimal Value { get; set; }
    public string Currency {get; set; }
}
```

This allows you to create the object and do whatever you want with it. The caller might put illegal values to the properties in such class. If you want to fix this, for example, adding logic to keep the Value positive, you need to watch the Value setter. But... what if the assigned value is negative? Throw an exception? Set Value to 0?

However, if this class is designed like this

```csharp
class Money 
{
    public Money(decimal value, string currency)
    {
          // check value validity
          // check currency validity
    }
    
    public decimal Value { get; }
    public string Currency { get; }
}
```

Whenever the calling code has successfully created and instance of Money, we know that instance is valid and can't turn into an invalid state.

Of course, mutability is necessary, but it's always a good idea to stick to immutability as much as we can. Also, immutable is friends with concurrency :)

**Reply by timunie** [reply]
Thank you so much for your explanation ❤️. That doesn't explain why an IObservable has a benefit over a normal (read-only) Property. You can have a private setter in there, so unless one is evil and uses reflection to edit the value, it should also be immutable. I'll update my proposal accordingly. 

Nevertheless, I'm happy you found the root cause of your issue and were able to solve it 👍

Happy coding
Tim

**Reply by SuperJMN** [reply]
Oh, I misunderstood then. When a value doesn't update at all, observables aren't useful. So, there's no benefit at all. 
Thanks Tim!

**Comment by SuperJMN** [comment]
Finally, I after days of research, I found the cause: a misuse of `System.Reactive`.

The reason why it didn't work is that the sequence emitted all the items before Avalonia subscribed to the observable. So, they were ignored and nothing happened.

Now, if you don't want subscribers to lose them, you need to use the `Replay` operator.

Please, look at the final code:

```csharp
public class MainWindowViewModel : ReactiveObject
{
    private readonly ReadOnlyObservableCollection<MyViewModel> myCollection;

    public MainWindowViewModel()
    {
        var changes = Observable.Defer(() => Observable.Range(1, 4))
            .Select(i => new MyViewModel(i))
            .ToObservableChangeSet()
            .ObserveOn(RxApp.MainThreadScheduler)
            .Replay()     // THIS!!
            .RefCount();

        changes
            .ObserveOn(RxApp.MainThreadScheduler)
            .Bind(out myCollection)
            .Subscribe();

        IsSomethingSelected = changes
            .AutoRefresh(x => x.IsSelected)
            .ToCollection()
            .Select(x => x.Any(model => model.IsSelected));
    }

    public IObservable<bool> IsSomethingSelected { get; }

    public ReadOnlyObservableCollection<MyViewModel> MyCollection => myCollection;
}

public class MyViewModel : ReactiveObject
{
    private bool isSelected;

    public MyViewModel(int i)
    {
        Number = i;
    }

    public int Number { get; }

    public bool IsSelected
    {
        get => isSelected;
        set => this.RaiseAndSetIfChanged(ref isSelected, value);
    }
}
```
